src/sgame/sg_bot_ai.cpp

```diff
@@ -1050,6 +1050,104 @@ AINodeStatus_t BotActionResetStuckTime( gentity_t *self, AIGenericNode_t* )
 	return AINodeStatus_t::STATUS_SUCCESS;
 }
 
+// resupply action does 2 things:
+// * it makes the bot move at range of a resupply building
+// * it makes the bot to actively reload when at range to
+//   prevent them to go back to fight with one less clip...
+//   or to be stuck in the action.
+// TODO: handle aliens hugs to regen ammo (that's true for
+//   the BotActionHeal, mind you). This likely requires support
+//   for crowds.
+AINodeStatus_t BotActionResupply( gentity_t *self, AIGenericNode_t *node )
+{
+	team_t team = G_Team(self);
+	playerState_t const& ps = self->client->ps;
+	weapon_t const wp = BG_PrimaryWeapon( ps.stats );
+	weaponAttributes_t const* wpa = BG_Weapon( wp );
+	ASSERT( self && self->botMind );
+	botMemory_t & mind = *self->botMind;
+	bool needsPoison = team == TEAM_ALIENS
+		&& !( ps.stats[ STAT_STATE ] & SS_POISONED )
+		&& mind.closestBuildings[ BA_A_BOOSTER ].ent;
+	bool needsAmmo = ps.ammo != wpa->maxAmmo || ps.clips != wpa->maxClips;
+
+	bool needsSupply = needsAmmo || needsPoison;
+
+	float range = 0.f; // poison requires touching the booster, hence 0.f val
+	if ( mind.currentNode != node )
+	{
+		// not sure if it's really logical to fail when the action
+		// is not needed, but imitating other actions (heal) here.
+		if ( !needsSupply )
+		{
+			return STATUS_FAILURE;
+		}
+
+		buildable_t dest;
+		switch( team )
+		{
+			case TEAM_ALIENS:
+				dest = BA_A_BOOSTER;
+				break;
+			case TEAM_HUMANS:
+				range = ENTITY_USE_RANGE;
+				if ( wpa->usesEnergy &&
+						mind.closestBuildings[ BA_H_REACTOR ].distance <= mind.closestBuildings[ BA_H_ARMOURY ].distance )
+				{
+					dest = BA_H_REACTOR;
+				}
+				else
+				{
+					dest = BA_H_ARMOURY;
+				}
+				break;
+			case TEAM_NONE:
+			case TEAM_ALL:
+			case NUM_TEAMS:
+				dest = BA_NONE;
```

Comment:
Why this code for an impossible case? Should be an assertion at most.

Reply:
I'm growing wary about assertions in unvanquished's case, seriously.

Reply:
If the bot is "playing" on spectators then something is severely broken and the assumptions of the program are violated everywhere.

If you are worried that this might happen, it would be better to add a check at the entry point of behavior tree evaluation, instead of in all the leaf nodes.